// crm-frontend/src/features/Mail/InternalMailPage.js
import React, { useState, useEffect, useRef, useMemo } from "react";
import { Link } from "react-router-dom";
import { 
  Mail, Edit3, Inbox, FileText, Send, Trash2, RefreshCw, 
  Search, Star, Paperclip, ChevronLeft, Download, Eye, 
  Trash, ArrowLeft, Loader2, CornerUpLeft, Plus, FolderOpen,
  HardDrive, SlidersHorizontal, Moon, Bell, ChevronDown, MoreVertical
} from "lucide-react";
import { internalMailApi } from "./internalMailApi";
import { getAppSocket } from "../../services/socket";
import axios from "axios";
import "./internalMail.css";

// Helper for consistent initials avatar backgrounds
const getAvatarBg = (name) => {
  const colors = [
    "#9b5de5", // Purple
    "#32d583", // Success Green
    "#57a6ff", // Info Blue
    "#f5b546", // Warning Amber
    "#e8420a", // Netcradus Orange
    "#f15bb5", // Pink
    "#00bbf9", // Cyan
    "#00f5d4"  // Mint
  ];
  let hash = 0;
  const str = name || "System Account";
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  return colors[Math.abs(hash) % colors.length];
};

export default function InternalMailPage() {
  const [currentFolder, setCurrentFolder] = useState("inbox"); // inbox, drafts, sent, deleted
  const [mails, setMails] = useState([]);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  
  // Counts
  const [inboxUnreadCount, setInboxUnreadCount] = useState(0);
  const [draftsCount, setDraftsCount] = useState(0);
  
  // Selection / Detail State
  const [selectedMailId, setSelectedMailId] = useState(null);
  const [selectedMail, setSelectedMail] = useState(null);
  const [loadingDetail, setLoadingDetail] = useState(false);

  // Search & Filter
  const [searchQuery, setSearchQuery] = useState("");
  const [activeFilter, setActiveFilter] = useState("all"); // all, unread, starred

  // Compose State
  const [composeMode, setComposeMode] = useState(false); // false, "new", "edit_draft"
  const [draftMailId, setDraftMailId] = useState(null);
  const [replyToId, setReplyToId] = useState(null);
  const [originalMailQuote, setOriginalMailQuote] = useState(null);
  
  // Compose Form Fields
  const [recipientsSearch, setRecipientsSearch] = useState("");
  const [selectedRecipients, setSelectedRecipients] = useState([]);
  const [subject, setSubject] = useState("");
  const [body, setBody] = useState("");
  const [files, setFiles] = useState([]);
  const [submitting, setSubmitting] = useState(false);

  // Directory for Compose
  const [userDirectory, setUserDirectory] = useState([]);
  const [showUserDropdown, setShowUserDropdown] = useState(false);

  // Mobile view
  const [mobileView, setMobileView] = useState("list"); // sidebar, list, detail

  const socketRef = useRef(null);
  const myUserId = localStorage.getItem("userId") || "";
  const myUserName = localStorage.getItem("userName") || "User";

  // Load user directory
  useEffect(() => {
    internalMailApi.getUsers()
      .then(res => {
        setUserDirectory(res.data?.data || []);
      })
      .catch(err => console.error("Failed to load user directory:", err));
  }, []);

  // Fetch mail count
  const fetchCounts = async () => {
    try {
      const unreadRes = await internalMailApi.getUnreadCount();
      setInboxUnreadCount(unreadRes.data?.data?.unreadCount || 0);

      const draftsRes = await internalMailApi.getDrafts();
      setDraftsCount(draftsRes.data?.data?.length || 0);
    } catch (err) {
      console.error("Failed to fetch mail counts:", err);
    }
  };

  // Fetch list of current folder
  const fetchFolderList = async (folder = currentFolder) => {
    try {
      setLoading(true);
      setError("");
      let res;
      if (folder === "inbox") res = await internalMailApi.getInbox();
      else if (folder === "sent") res = await internalMailApi.getSent();
      else if (folder === "drafts") res = await internalMailApi.getDrafts();
      else if (folder === "deleted") res = await internalMailApi.getDeleted();

      setMails(res.data?.data || []);
      fetchCounts();
    } catch (err) {
      setError("Failed to load mails.");
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchFolderList(currentFolder);
    setComposeMode(false);
    setSelectedMailId(null);
    setSelectedMail(null);
  }, [currentFolder]);

  // Load single mail detail
  useEffect(() => {
    if (selectedMailId) {
      setLoadingDetail(true);
      internalMailApi.getMail(selectedMailId)
        .then(res => {
          setSelectedMail(res.data?.data || null);
          if (currentFolder === "inbox") {
            setMails(prev => prev.map(m => m._id === selectedMailId ? { ...m, readBy: [...(m.readBy || []), myUserId] } : m));
            fetchCounts();
          }
        })
        .catch(() => {
          setError("Failed to load email detail.");
        })
        .finally(() => {
          setLoadingDetail(false);
        });
    }
  }, [selectedMailId]);

  // Socket.IO sync
  useEffect(() => {
    const token = localStorage.getItem("token");
    const socket = getAppSocket(token);
    socketRef.current = socket;

    if (socket) {
      const handleNewMail = () => {
        fetchFolderList();
      };
      socket.on("new_mail", handleNewMail);

      return () => {
        socket.off("new_mail", handleNewMail);
      };
    }
  }, [currentFolder]);

  // Filter and search
  const filteredMails = useMemo(() => {
    let list = mails;

    if (activeFilter === "starred") {
      list = list.filter(m => m.starredBy?.includes(myUserId));
    } else if (activeFilter === "unread") {
      list = list.filter(m => !m.readBy?.includes(myUserId));
    }

    if (searchQuery.trim()) {
      const q = searchQuery.toLowerCase().trim();
      list = list.filter(m => {
        const senderName = m.sender?.name || m.sender?.fullName || "";
        const subjectStr = m.subject || "";
        const bodyStr = m.body || "";
        return senderName.toLowerCase().includes(q) || 
               subjectStr.toLowerCase().includes(q) || 
               bodyStr.toLowerCase().includes(q);
      });
    }

    return list;
  }, [mails, activeFilter, searchQuery, myUserId]);

  // User Picker
  const filteredUserDirectory = useMemo(() => {
    if (!recipientsSearch.trim()) return [];
    const q = recipientsSearch.toLowerCase().trim();
    return userDirectory.filter(u => 
      u._id !== myUserId &&
      !selectedRecipients.some(sel => sel._id === u._id) &&
      ((u.name || "").toLowerCase().includes(q) || (u.email || "").toLowerCase().includes(q))
    );
  }, [recipientsSearch, userDirectory, selectedRecipients, myUserId]);

  // Toggle star
  const handleToggleStar = async (e, mailId, currentStarred) => {
    e.stopPropagation();
    try {
      const nextStar = !currentStarred;
      const res = await internalMailApi.toggleStar(mailId, nextStar);
      if (res.data?.success) {
        setMails(prev => prev.map(m => {
          if (m._id === mailId) {
            const starredBy = m.starredBy || [];
            return {
              ...m,
              starredBy: nextStar 
                ? [...starredBy, myUserId] 
                : starredBy.filter(id => id !== myUserId)
            };
          }
          return m;
        }));

        if (selectedMail && selectedMail._id === mailId) {
          const starredBy = selectedMail.starredBy || [];
          setSelectedMail(prev => ({
            ...prev,
            starredBy: nextStar 
              ? [...starredBy, myUserId] 
              : starredBy.filter(id => id !== myUserId)
          }));
        }
      }
    } catch (err) {
      console.error("Star toggle failed:", err);
    }
  };

  // Delete
  const handleDeleteMail = async (mailId) => {
    try {
      const res = await internalMailApi.softDelete(mailId);
      if (res.data?.success) {
        setMails(prev => prev.filter(m => m._id !== mailId));
        setSelectedMail(null);
        setSelectedMailId(null);
        fetchCounts();
        if (mobileView === "detail") setMobileView("list");
      }
    } catch (err) {
      alert("Failed to delete email.");
    }
  };

  // Restore
  const handleRestoreMail = async (mailId) => {
    try {
      const res = await internalMailApi.restoreMail(mailId);
      if (res.data?.success) {
        setMails(prev => prev.filter(m => m._id !== mailId));
        setSelectedMail(null);
        setSelectedMailId(null);
        fetchCounts();
        if (mobileView === "detail") setMobileView("list");
      }
    } catch (err) {
      alert("Failed to restore email.");
    }
  };

  // Download attachment
  const handleDownloadAttachment = async (attachmentId, originalname) => {
    try {
      const url = internalMailApi.getDownloadUrl(selectedMail._id, attachmentId);
      const res = await axios.get(url, {
        headers: { Authorization: `Bearer ${localStorage.getItem("token")}` },
        responseType: "blob"
      });
      const blobUrl = window.URL.createObjectURL(new Blob([res.data]));
      const link = document.createElement("a");
      link.href = blobUrl;
      link.setAttribute("download", originalname);
      document.body.appendChild(link);
      link.click();
      link.remove();
    } catch (err) {
      alert("Failed to download attachment.");
    }
  };

  // View attachment inline
  const handleViewAttachment = (attachmentId) => {
    const url = internalMailApi.getDownloadUrl(selectedMail._id, attachmentId);
    window.open(url, "_blank");
  };

  // Compose / Reply Triggers
  const handleComposeNew = () => {
    setComposeMode("new");
    setDraftMailId(null);
    setReplyToId(null);
    setOriginalMailQuote(null);
    setSelectedRecipients([]);
    setSubject("");
    setBody("");
    setFiles([]);
    setMobileView("detail");
  };

  const handleReply = (original) => {
    setComposeMode("new");
    setDraftMailId(null);
    setReplyToId(original._id);
    setOriginalMailQuote(original);
    
    const senderObj = original.sender && original.sender._id 
      ? original.sender 
      : { _id: original.sender, name: original.sender?.fullName || original.sender?.name || "System" };

    setSelectedRecipients([senderObj]);
    
    // Subject Re prefix checks
    const originalSubject = original.subject || "";
    const startsWithRe = /^Re:\s*/i.test(originalSubject);
    setSubject(startsWithRe ? originalSubject : `Re: ${originalSubject}`);
    
    setBody("");
    setFiles([]);
    setMobileView("detail");
  };

  const handleEditDraft = (draft) => {
    setComposeMode("edit_draft");
    setDraftMailId(draft._id);
    setSubject(draft.subject || "");
    setBody(draft.body || "");
    setFiles([]);

    if (draft.replyTo) {
      setReplyToId(draft.replyTo._id || draft.replyTo);
      setOriginalMailQuote(draft.replyTo);
    } else {
      setReplyToId(null);
      setOriginalMailQuote(null);
    }
    
    const resolvedRecs = (draft.recipients || []).map(r => {
      const found = userDirectory.find(u => u._id === (r._id || r));
      return found || { _id: (r._id || r), name: r.fullName || r.name || "User", email: r.email || "" };
    });
    setSelectedRecipients(resolvedRecs);
    setMobileView("detail");
  };

  const handleFileChange = (e) => {
    const chosenFiles = Array.from(e.target.files);
    setFiles(prev => [...prev, ...chosenFiles].slice(0, 5));
  };

  const handleRemoveFile = (index) => {
    setFiles(prev => prev.filter((_, idx) => idx !== index));
  };

  const handleSubmitMail = async (e, shouldSend = true) => {
    e.preventDefault();
    if (shouldSend && selectedRecipients.length === 0) {
      alert("Please select at least one recipient.");
      return;
    }

    try {
      setSubmitting(true);
      const formData = new FormData();
      selectedRecipients.forEach(r => formData.append("recipients", r._id));
      formData.append("subject", subject);
      formData.append("body", body);
      files.forEach(f => formData.append("files", f));

      if (replyToId) {
        formData.append("replyTo", replyToId);
      }

      let res;
      if (shouldSend) {
        if (composeMode === "edit_draft" && draftMailId) {
          formData.append("draftId", draftMailId);
        }
        res = await internalMailApi.sendMail(formData);
      } else {
        if (composeMode === "edit_draft" && draftMailId) {
          formData.append("status", "draft");
          res = await internalMailApi.updateDraft(draftMailId, formData);
        } else {
          res = await internalMailApi.saveDraft(formData);
        }
      }

      if (res.data?.success) {
        setComposeMode(false);
        setDraftMailId(null);
        setSelectedRecipients([]);
        setSubject("");
        setBody("");
        setFiles([]);
        fetchFolderList(currentFolder);
        if (mobileView === "detail") setMobileView("list");
      } else {
        alert(res.data?.message || "Operation failed.");
      }
    } catch (err) {
      alert(err.response?.data?.message || "An error occurred.");
    } finally {
      setSubmitting(false);
    }
  };

  const formatMailDate = (value) => {
    if (!value) return "";
    const date = new Date(value);
    const today = new Date();
    if (date.toDateString() === today.toDateString()) {
      return date.toLocaleTimeString("en-IN", { hour: "2-digit", minute: "2-digit" });
    }
    return date.toLocaleDateString("en-IN", { day: "2-digit", month: "short" });
  };

  const getInitials = (name) => {
    if (!name) return "S";
    return name.slice(0, 2).toUpperCase();
  };

  return (
    <div className="netmail-container">
      
      {/* 1. CUSTOM TOPBAR HEADER */}
      <header className="netmail-header">
        <div style={{ display: "flex", alignItems: "center", gap: "24px" }}>
          <img src="/mailLogo.png" alt="NETMAIL" style={{ height: "20px", objectFit: "contain" }} />
          <Link to="/dashboard" style={{ display: "flex", alignItems: "center", gap: "6px", fontSize: "12px", color: "var(--color-text-secondary)", textDecoration: "none", fontWeight: "600" }}>
            <ArrowLeft size={14} /> Back to CRM
          </Link>
        </div>

        <div className="netmail-global-search">
          <Search size={14} style={{ opacity: 0.6 }} />
          <input 
            placeholder="Search emails, contacts or subject..." 
            value={searchQuery} 
            onChange={e => setSearchQuery(e.target.value)} 
          />
          <span style={{ fontSize: "10px", opacity: 0.5, padding: "2px 6px", border: "1px solid rgba(255,255,255,0.1)", borderRadius: "4px", fontWeight: "bold" }}>⌘K</span>
        </div>

        <div style={{ display: "flex", alignItems: "center", gap: "20px" }}>
          <button className="netmail-action-btn" title="Theme toggler">
            <Moon size={18} />
          </button>
          
          <div style={{ position: "relative", cursor: "pointer" }} onClick={() => setCurrentFolder("inbox")}>
            <Mail size={18} style={{ opacity: 0.8 }} />
            {inboxUnreadCount > 0 && (
              <span style={{ position: "absolute", top: "-5px", right: "-5px", backgroundColor: "#ff6363", color: "#fff", fontSize: "9px", fontWeight: "bold", borderRadius: "50%", minWidth: "14px", height: "14px", display: "flex", alignItems: "center", justifyContent: "center" }}>
                {inboxUnreadCount}
              </span>
            )}
          </div>

          <div style={{ position: "relative", cursor: "pointer" }}>
            <Bell size={18} style={{ opacity: 0.8 }} />
            <span style={{ position: "absolute", top: "-5px", right: "-5px", backgroundColor: "#e8420a", color: "#fff", fontSize: "9px", fontWeight: "bold", borderRadius: "50%", minWidth: "14px", height: "14px", display: "flex", alignItems: "center", justifyContent: "center" }}>
              12
            </span>
          </div>

          <div style={{ display: "flex", alignItems: "center", gap: "8px", cursor: "pointer", borderLeft: "1px solid rgba(255,255,255,0.08)", paddingLeft: "16px" }}>
            <div style={{ width: "28px", height: "28px", borderRadius: "50%", backgroundColor: "#e8420a", color: "#fff", display: "flex", alignItems: "center", justifyContent: "center", fontSize: "11px", fontWeight: "bold" }}>
              {getInitials(myUserName)}
            </div>
            <span style={{ fontSize: "12px", fontWeight: "600", display: "flex", alignItems: "center", gap: "4px" }}>
              {myUserName} <ChevronDown size={12} style={{ opacity: 0.6 }} />
            </span>
          </div>
        </div>
      </header>

      {/* 2. THREE COLUMN LAYOUT */}
      <div className="netmail-layout">
        
        {/* SIDEBAR */}
        <aside className={`netmail-sidebar ${mobileView === "sidebar" ? "active-mobile" : ""}`}>
          <button type="button" className="netmail-compose-btn" onClick={handleComposeNew}>
            <Plus size={16} />
            <span>Compose</span>
          </button>

          <nav className="netmail-menu">
            <div className={`netmail-menu-item ${currentFolder === "inbox" ? "active" : ""}`} onClick={() => { setCurrentFolder("inbox"); setMobileView("list"); }}>
              <span style={{ display: "flex", alignItems: "center", gap: "12px" }}><Inbox size={16} /> <span>Inbox</span></span>
              {inboxUnreadCount > 0 && <span className="netmail-badge">{inboxUnreadCount}</span>}
            </div>

            <div className={`netmail-menu-item ${currentFolder === "drafts" ? "active" : ""}`} onClick={() => { setCurrentFolder("drafts"); setMobileView("list"); }}>
              <span style={{ display: "flex", alignItems: "center", gap: "12px" }}><FileText size={16} /> <span>Drafts</span></span>
              {draftsCount > 0 && <span className="netmail-badge">{draftsCount}</span>}
            </div>

            <div className={`netmail-menu-item ${currentFolder === "sent" ? "active" : ""}`} onClick={() => { setCurrentFolder("sent"); setMobileView("list"); }}>
              <span style={{ display: "flex", alignItems: "center", gap: "12px" }}><Send size={16} /> <span>Sent</span></span>
            </div>

            <div className={`netmail-menu-item ${currentFolder === "deleted" ? "active" : ""}`} onClick={() => { setCurrentFolder("deleted"); setMobileView("list"); }}>
              <span style={{ display: "flex", alignItems: "center", gap: "12px" }}><Trash2 size={16} /> <span>Deleted</span></span>
            </div>
          </nav>

          <div className="netmail-sidebar-storage">
            <div style={{ display: "flex", alignItems: "center", gap: "8px", fontSize: "12px", fontWeight: "600", color: "var(--color-text-secondary)" }}>
              <HardDrive size={14} style={{ opacity: 0.6 }} />
              <span>Storage</span>
            </div>
            <div className="netmail-progress-bar-bg">
              <div className="netmail-progress-bar-fill" style={{ width: "12.4%" }}></div>
            </div>
            <div style={{ fontSize: "11px", color: "var(--color-text-muted)" }}>1.24 GB of 10 GB used</div>
          </div>
        </aside>

        {/* MIDDLE COLUMN: Threads list */}
        <section className={`netmail-list-col ${mobileView === "list" ? "active-mobile" : ""}`}>
          <div className="netmail-list-header">
            <div className="netmail-list-toolbar">
              <button className="netmail-filter-btn md-hidden" onClick={() => setMobileView("sidebar")} style={{ display: "flex", alignItems: "center", gap: "4px" }}>
                <ChevronLeft size={14} /> Menu
              </button>
              <h3 style={{ textTransform: "capitalize" }}>{currentFolder} {currentFolder === "inbox" && inboxUnreadCount > 0 ? `(${inboxUnreadCount})` : ""}</h3>
              <div style={{ display: "flex", gap: "4px" }}>
                <button type="button" className="netmail-action-btn" onClick={() => fetchFolderList(currentFolder)} title="Refresh folder">
                  <RefreshCw size={14} />
                </button>
                <button type="button" className="netmail-action-btn" title="Filters">
                  <SlidersHorizontal size={14} />
                </button>
              </div>
            </div>

            <div className="netmail-inner-search">
              <Search size={14} style={{ opacity: 0.6 }} />
              <input 
                placeholder={`Search in ${currentFolder}...`}
                value={searchQuery}
                onChange={e => setSearchQuery(e.target.value)}
              />
            </div>

            <div className="netmail-filters">
              <button className={`netmail-filter-chip ${activeFilter === "all" ? "active" : ""}`} onClick={() => setActiveFilter("all")}>All</button>
              <button className={`netmail-filter-chip ${activeFilter === "unread" ? "active" : ""}`} onClick={() => setActiveFilter("unread")}>Unread</button>
              <button className={`netmail-filter-chip ${activeFilter === "starred" ? "active" : ""}`} onClick={() => setActiveFilter("starred")}>Starred</button>
            </div>
          </div>

          <div className="netmail-list-scroll">
            {loading ? (
              <div style={{ display: "flex", flexDirection: "column", gap: "8px", padding: "60px 40px", alignItems: "center", color: "var(--color-text-muted)" }}>
                <Loader2 className="animate-spin" size={24} color="var(--color-accent)" />
                <span style={{ fontSize: "12px" }}>Synchronizing...</span>
              </div>
            ) : filteredMails.length === 0 ? (
              <div className="netmail-empty-box">
                <FolderOpen size={36} style={{ opacity: 0.2, color: "var(--color-text-muted)" }} />
                <span className="netmail-empty-title">
                  {searchQuery ? "No search results" : `No emails in ${currentFolder}`}
                </span>
                <span className="netmail-empty-sub">
                  {searchQuery 
                    ? `No email matches your query "${searchQuery}"`
                    : `This folder does not contain any messages.`}
                </span>
              </div>
            ) : (
              filteredMails.map((item) => {
                const isRead = item.readBy?.includes(myUserId);
                const isStarred = item.starredBy?.includes(myUserId);
                const hasAttachment = item.attachments && item.attachments.length > 0;

                // Sender display resolution
                let displayUser = "System Account";
                let initialsSource = "System Account";

                if (currentFolder === "sent") {
                  const recNames = (item.recipients || []).map(r => r.fullName || r.name || "User").join(", ");
                  displayUser = recNames ? `To: ${recNames}` : "To: Drafts";
                  initialsSource = recNames || "Drafts";
                } else if (currentFolder === "drafts") {
                  const recNames = (item.recipients || []).map(r => r.fullName || r.name || "User").join(", ");
                  displayUser = recNames ? `To: ${recNames}` : "(No Recipient)";
                  initialsSource = recNames || "Drafts";
                } else {
                  displayUser = item.sender?._id === myUserId ? "Me" : (item.sender?.fullName || item.sender?.name || "System Account");
                  initialsSource = item.sender?.fullName || item.sender?.name || "System Account";
                }

                return (
                  <div 
                    key={item._id} 
                    className={`netmail-row-card ${selectedMailId === item._id ? "selected" : ""} ${(!isRead && currentFolder === "inbox") ? "unread" : ""}`}
                    onClick={() => {
                      if (currentFolder === "drafts") {
                        handleEditDraft(item);
                      } else {
                        setComposeMode(false);
                        setSelectedMailId(item._id);
                      }
                      setMobileView("detail");
                    }}
                  >
                    {/* Unread small orange dot to the left of the avatar */}
                    {!isRead && currentFolder === "inbox" && (
                      <span style={{ width: "6px", height: "6px", borderRadius: "50%", backgroundColor: "var(--color-accent)", position: "absolute", left: "8px", top: "50%", transform: "translateY(-50%)" }}></span>
                    )}

                    <div className="netmail-row-avatar" style={{ backgroundColor: getAvatarBg(initialsSource) }}>
                      {getInitials(initialsSource)}
                    </div>

                    <div className="netmail-row-details">
                      <div className="netmail-row-meta">
                        <span className="netmail-row-sender">{displayUser}</span>
                        <span className="netmail-row-date">{formatMailDate(item.createdAt)}</span>
                      </div>
                      <div className="netmail-row-subject">
                        {currentFolder === "drafts" && <span className="netmail-draft-badge">Draft</span>}
                        <span>{item.subject || "(No Subject)"}</span>
                      </div>
                      <div className="netmail-row-preview">{item.body || "(No message body)"}</div>
                      
                      <div className="netmail-row-indicators">
                        <div>
                          {hasAttachment && <Paperclip size={12} style={{ opacity: 0.5, color: "var(--color-text-muted)" }} />}
                        </div>
                        <Star 
                          size={13} 
                          className={`netmail-star-icon ${isStarred ? "starred" : ""}`} 
                          onClick={(e) => handleToggleStar(e, item._id, isStarred)} 
                        />
                      </div>
                    </div>
                  </div>
                );
              })
            )}
          </div>
        </section>

        {/* DETAIL VIEW / COMPOSE WINDOW */}
        <main className={`netmail-detail-col ${mobileView === "detail" ? "active-mobile" : ""}`}>
          
          {/* Mobile Back navigation */}
          <div className="netmail-list-toolbar md-hidden" style={{ padding: "8px 16px", borderBottom: "1px solid rgba(255,255,255,0.03)", minHeight: "44px" }}>
            <button className="netmail-filter-btn" onClick={() => setMobileView("list")} style={{ display: "flex", alignItems: "center", gap: "4px" }}>
              <ChevronLeft size={14} /> Back to list
            </button>
          </div>

          {composeMode ? (
            /* COMPOSE / REPLY MODE */
            <div className="netmail-compose">
              <div className="netmail-compose-header">
                {replyToId ? "Reply to Mail" : composeMode === "new" ? "New Mail" : "Edit Draft"}
              </div>

              <form onSubmit={(e) => handleSubmitMail(e, true)} className="netmail-compose-form">
                <div className="netmail-form-row">
                  <label>To</label>
                  <div className="netmail-recipient-picker-wrapper">
                    {selectedRecipients.map((rec) => (
                      <span key={rec._id} className="netmail-recipient-bubble">
                        {rec.fullName || rec.name}
                        {!replyToId && (
                          <button type="button" onClick={() => setSelectedRecipients(prev => prev.filter(p => p._id !== rec._id))} style={{ background: "none", border: "none", color: "var(--color-accent)", cursor: "pointer", marginLeft: "4px", fontSize: "12px", fontWeight: "bold" }}>
                            ×
                          </button>
                        )}
                      </span>
                    ))}
                    {!replyToId && (
                      <input
                        type="text"
                        className="netmail-recipient-input"
                        placeholder={selectedRecipients.length === 0 ? "Search internal CRM users..." : ""}
                        value={recipientsSearch}
                        onChange={e => {
                          setRecipientsSearch(e.target.value);
                          setShowUserDropdown(true);
                        }}
                        onFocus={() => setShowUserDropdown(true)}
                      />
                    )}
                  </div>

                  {showUserDropdown && filteredUserDirectory.length > 0 && (
                    <div className="netmail-user-dropdown">
                      {filteredUserDirectory.map((user) => (
                        <div 
                          key={user._id} 
                          className="netmail-dropdown-item"
                          onClick={() => {
                            setSelectedRecipients(prev => [...prev, user]);
                            setRecipientsSearch("");
                            setShowUserDropdown(false);
                          }}
                        >
                          <span style={{ fontWeight: "700", color: "#fff" }}>{user.fullName || user.name}</span>
                          <span style={{ color: "var(--color-text-muted)" }}>{user.email} ({user.role} | {user.department || "General"})</span>
                        </div>
                      ))}
                    </div>
                  )}
                </div>

                <div className="netmail-form-row">
                  <label>Subject</label>
                  <input
                    type="text"
                    className="netmail-form-input"
                    placeholder="Subject line"
                    value={subject}
                    onChange={e => setSubject(e.target.value)}
                    maxLength={200}
                  />
                </div>

                <div className="netmail-form-row" style={{ flexGrow: 1 }}>
                  <label>Message</label>
                  <textarea
                    className="netmail-form-textarea"
                    placeholder="Write your email message body..."
                    value={body}
                    onChange={e => setBody(e.target.value)}
                    maxLength={10000}
                    style={{ height: "100%", minHeight: "160px" }}
                  />
                  {originalMailQuote && (
                    /* Separated quoted reference section */
                    <div style={{ 
                      marginTop: "16px", 
                      padding: "12px 16px", 
                      borderTop: "1px dashed rgba(255, 255, 255, 0.08)",
                      backgroundColor: "rgba(255, 255, 255, 0.01)",
                      borderRadius: "8px" 
                    }}>
                      <div style={{ fontSize: "11px", color: "var(--color-text-muted)", marginBottom: "6px" }}>
                        On {new Date(originalMailQuote.createdAt).toLocaleString("en-IN")} at {originalMailQuote.sender?.fullName || originalMailQuote.sender?.name || "System"} wrote:
                      </div>
                      <blockquote style={{ 
                        margin: 0, 
                        paddingLeft: "12px", 
                        borderLeft: "2px solid var(--color-accent)", 
                        color: "var(--color-text-secondary)",
                        fontSize: "13px",
                        whiteSpace: "pre-wrap",
                        fontStyle: "italic",
                        lineHeight: "1.5"
                      }}>
                        {originalMailQuote.body || "(No message body)"}
                      </blockquote>
                    </div>
                  )}
                </div>

                {/* Upload Slot */}
                <div className="netmail-form-row">
                  <label style={{ display: "flex", alignItems: "center", gap: "6px", cursor: "pointer", fontSize: "12px", width: "fit-content" }} className="hover-text-primary">
                    <Paperclip size={14} color="var(--color-accent)" /> 
                    <span style={{ fontWeight: "700" }}>Attach File</span> (Max 10MB)
                    <input
                      type="file"
                      multiple
                      onChange={handleFileChange}
                      style={{ display: "none" }}
                    />
                  </label>

                  {files.length > 0 && (
                    <div className="netmail-attachment-preview-list" style={{ marginTop: "8px" }}>
                      {files.map((file, idx) => (
                        <div key={idx} className="netmail-attachment-preview-card" style={{ display: "flex", justifyContent: "space-between", alignItems: "center", padding: "6px 12px", border: "1px solid rgba(255,255,255,0.08)", borderRadius: "6px", background: "rgba(255,255,255,0.01)" }}>
                          <span style={{ overflow: "hidden", textOverflow: "ellipsis", whiteSpace: "nowrap", fontSize: "12px" }}>
                            {file.name} ({(file.size / (1024 * 1024)).toFixed(2)} MB)
                          </span>
                          <button type="button" className="btn-icon" onClick={() => handleRemoveFile(idx)} style={{ padding: 0 }}>
                            <Trash size={12} color="var(--color-error)" />
                          </button>
                        </div>
                      ))}
                    </div>
                  )}
                </div>

                <div className="netmail-compose-footer">
                  <div style={{ display: "flex", gap: "8px" }}>
                    <button type="submit" className="btn btn-primary" disabled={submitting}>
                      {submitting ? "Sending..." : "Send Mail"}
                    </button>
                    <button type="button" className="btn btn-secondary" disabled={submitting} onClick={(e) => handleSubmitMail(e, false)}>
                      Save as Draft
                    </button>
                  </div>
                  <button type="button" className="btn btn-ghost" onClick={() => { setComposeMode(false); if (mobileView === "detail") setMobileView("list"); }}>
                    Cancel
                  </button>
                </div>
              </form>
            </div>
          ) : selectedMailId && selectedMail ? (
            /* MESSAGE READER PANEL */
            (() => {
              const showReplyButton = 
                currentFolder === "inbox" && 
                selectedMail.sender && 
                String(selectedMail.sender._id || selectedMail.sender) !== myUserId &&
                selectedMail.status !== "draft" &&
                !selectedMail.deletedBy?.includes(myUserId);

              const initialsSource = selectedMail.sender?.fullName || selectedMail.sender?.name || "System Account";

              return (
                <div style={{ display: "flex", flexDirection: "column", height: "100%", overflow: "hidden" }}>
                  
                  {/* Toolbar Actions matching reference */}
                  <div className="netmail-detail-toolbar">
                    <div className="netmail-detail-toolbar-left">
                      {showReplyButton && (
                        <button 
                          type="button" 
                          className="netmail-action-btn-primary" 
                          onClick={() => handleReply(selectedMail)}
                          title="Reply to this mail"
                          aria-label="Reply to this mail"
                        >
                          <CornerUpLeft size={14} />
                          <span>Reply</span>
                        </button>
                      )}

                      <button 
                        type="button"
                        className="netmail-action-btn"
                        onClick={(e) => handleToggleStar(e, selectedMail._id, selectedMail.starredBy?.includes(myUserId))}
                        title="Star"
                      >
                        <Star 
                          size={16} 
                          className={selectedMail.starredBy?.includes(myUserId) ? "starred" : ""} 
                        />
                      </button>

                      {currentFolder === "deleted" ? (
                        <button 
                          type="button" 
                          className="netmail-action-btn" 
                          onClick={() => handleRestoreMail(selectedMail._id)}
                          title="Restore Mail"
                        >
                          <RefreshCw size={14} />
                        </button>
                      ) : (
                        <button 
                          type="button" 
                          className="netmail-action-btn" 
                          onClick={() => handleDeleteMail(selectedMail._id)}
                          title="Delete"
                        >
                          <Trash2 size={16} />
                        </button>
                      )}
                    </div>
                    
                    <div className="netmail-detail-toolbar-right">
                      <button className="netmail-action-btn" title="More options">
                        <MoreVertical size={16} />
                      </button>
                    </div>
                  </div>

                  {/* Mail scroll panel details */}
                  <div className="netmail-reader-scroll">
                    <div className="netmail-detail-subject-row">
                      <h2 className="netmail-detail-subject">{selectedMail.subject || "(No Subject)"}</h2>
                      <span style={{ fontSize: "10px", textTransform: "uppercase", padding: "2px 6px", border: "1px solid rgba(255,255,255,0.1)", borderRadius: "4px", color: "var(--color-text-muted)" }}>{currentFolder}</span>
                    </div>

                    <div className="netmail-meta-header">
                      <div className="netmail-sender-block">
                        <div className="netmail-sender-avatar" style={{ backgroundColor: getAvatarBg(initialsSource) }}>
                          {getInitials(initialsSource)}
                        </div>
                        <div className="netmail-sender-info">
                          <div className="netmail-sender-name-line">
                            <strong>{selectedMail.sender?.fullName || selectedMail.sender?.name || "System Account"}</strong>
                            {selectedMail.sender?.email && <span style={{ opacity: 0.6 }}>&lt;{selectedMail.sender.email}&gt;</span>}
                          </div>
                          <div className="netmail-recipient-line">
                            To: {selectedMail.recipients?.map(r => r.fullName || r.name || "User").join(", ")}
                          </div>
                        </div>
                      </div>

                      <div className="netmail-detail-date-col">
                        <span>
                          {new Date(selectedMail.createdAt).toLocaleString("en-IN", {
                            day: "2-digit",
                            month: "short",
                            year: "numeric",
                            hour: "2-digit",
                            minute: "2-digit"
                          })}
                        </span>
                        
                        <div style={{ display: "flex", gap: "6px" }}>
                          <Star 
                            size={14} 
                            style={{ cursor: "pointer" }}
                            className={selectedMail.starredBy?.includes(myUserId) ? "starred" : ""}
                            onClick={(e) => handleToggleStar(e, selectedMail._id, selectedMail.starredBy?.includes(myUserId))}
                          />
                          {showReplyButton && (
                            <CornerUpLeft 
                              size={14} 
                              style={{ cursor: "pointer" }}
                              onClick={() => handleReply(selectedMail)}
                            />
                          )}
                          <MoreVertical size={14} style={{ cursor: "pointer", opacity: 0.6 }} />
                        </div>
                      </div>
                    </div>

                    <div className="netmail-body-content">
                      {selectedMail.body || <span style={{ fontStyle: "italic", opacity: 0.4 }}>(No message content body)</span>}
                    </div>

                    {/* Collapsed/Visual Quoted Thread context */}
                    {selectedMail.replyTo && (
                      <div style={{ 
                        marginTop: "24px", 
                        padding: "16px", 
                        borderTop: "1px dashed rgba(255, 255, 255, 0.08)",
                        backgroundColor: "rgba(255, 255, 255, 0.01)",
                        borderRadius: "8px",
                        maxWidth: "820px"
                      }}>
                        <div style={{ fontSize: "11px", color: "var(--color-text-muted)", marginBottom: "6px", fontWeight: "700" }}>
                          --- Quoted Message ---
                        </div>
                        <div style={{ fontSize: "11px", color: "var(--color-text-muted)", marginBottom: "4px" }}>
                          From: {selectedMail.replyTo.sender?.fullName || selectedMail.replyTo.sender?.name || "System Account"}
                        </div>
                        <blockquote style={{ 
                          margin: 0, 
                          paddingLeft: "12px", 
                          borderLeft: "2px solid var(--color-accent)", 
                          color: "var(--color-text-secondary)",
                          fontSize: "13px",
                          whiteSpace: "pre-wrap",
                          fontStyle: "italic",
                          lineHeight: "1.6"
                        }}>
                          {selectedMail.replyTo.body || "(No message body)"}
                        </blockquote>
                      </div>
                    )}

                    {/* Premium Attachments Section */}
                    {selectedMail.attachments && selectedMail.attachments.length > 0 && (
                      <div className="netmail-attachments-container">
                        <div className="netmail-attachments-header">
                          {selectedMail.attachments.length} {selectedMail.attachments.length === 1 ? "Attachment" : "Attachments"}
                        </div>
                        <div className="netmail-attachments-grid">
                          {selectedMail.attachments.map((att) => {
                            const ext = att.originalname?.split(".").pop().toUpperCase() || "FILE";
                            return (
                              <div key={att._id} className="netmail-attachment-tile">
                                <div className="netmail-file-icon">
                                  {ext}
                                </div>
                                <div className="netmail-attachment-detail">
                                  <span className="netmail-attachment-filename" title={att.originalname}>{att.originalname}</span>
                                  <span className="netmail-attachment-size">{(att.size / 1024).toFixed(1)} KB</span>
                                </div>
                                <div className="netmail-attachment-actions">
                                  <button 
                                    type="button" 
                                    className="netmail-action-btn" 
                                    onClick={() => handleDownloadAttachment(att._id, att.originalname)}
                                    title="Download"
                                  >
                                    <Download size={14} />
                                  </button>
                                  <button 
                                    type="button" 
                                    className="netmail-action-btn" 
                                    onClick={() => handleViewAttachment(att._id)}
                                    title="View"
                                  >
                                    <Eye size={14} />
                                  </button>
                                </div>
                              </div>
                            );
                          })}
                        </div>
                      </div>
                    )}

                    {/* Bottom Quick Reply Trigger Bar */}
                    {showReplyButton && (
                      <div 
                        onClick={() => handleReply(selectedMail)} 
                        style={{ 
                          marginTop: "32px", 
                          padding: "16px 20px", 
                          border: "1px solid rgba(255, 255, 255, 0.08)", 
                          borderRadius: "12px", 
                          cursor: "pointer", 
                          display: "flex", 
                          justifyContent: "space-between", 
                          alignItems: "center",
                          color: "var(--color-text-muted)",
                          backgroundColor: "rgba(255, 255, 255, 0.015)",
                          maxWidth: "820px",
                          transition: "all 0.2s ease"
                        }}
                        onMouseEnter={e => {
                          e.currentTarget.style.backgroundColor = "rgba(255,255,255,0.03)";
                          e.currentTarget.style.borderColor = "rgba(255,255,255,0.12)";
                        }}
                        onMouseLeave={e => {
                          e.currentTarget.style.backgroundColor = "rgba(255,255,255,0.015)";
                          e.currentTarget.style.borderColor = "rgba(255,255,255,0.08)";
                        }}
                      >
                        <div style={{ display: "flex", alignItems: "center", gap: "10px" }}>
                          <div style={{ width: "24px", height: "24px", borderRadius: "50%", backgroundColor: "rgba(255,255,255,0.08)", color: "#fff", display: "flex", alignItems: "center", justifyContent: "center", fontSize: "10px", fontWeight: "bold" }}>
                            {getInitials(myUserName)}
                          </div>
                          <span style={{ fontSize: "13px" }}>Click here to Reply or Forward...</span>
                        </div>
                        <CornerUpLeft size={14} style={{ opacity: 0.6 }} />
                      </div>
                    )}
                  </div>
                </div>
              );
            })()
          ) : loadingDetail ? (
            <div style={{ display: "flex", flexDirection: "column", gap: "10px", padding: "100px", alignItems: "center", color: "var(--color-text-muted)" }}>
              <Loader2 className="animate-spin" size={28} color="var(--color-accent)" />
              <span style={{ fontSize: "12px" }}>Opening thread...</span>
            </div>
          ) : (
            /* NO SELECTED MAIL DEFAULT PANELS */
            <div className="netmail-empty-box">
              <div style={{ 
                width: "60px", 
                height: "60px", 
                borderRadius: "50%", 
                backgroundColor: "rgba(255,255,255,0.02)", 
                display: "flex", 
                alignItems: "center", 
                justifyContent: "center",
                marginBottom: "8px",
                border: "1px solid rgba(255,255,255,0.04)"
              }}>
                <Mail size={24} style={{ opacity: 0.3, color: "var(--color-text-secondary)" }} />
              </div>
              <span className="netmail-empty-title">Select an email to read</span>
              <span className="netmail-empty-sub">Choose a conversation from the folder list to see detailed messages.</span>
            </div>
          )}

        </main>

      </div>
    </div>
  );
}
